library/core/src/mem/transmutability.rs

```diff
@@ -13,6 +66,53 @@ pub unsafe trait BikeshedIntrinsicFrom<Src, const ASSUME: Assume = { Assume::NOT
 where
     Src: ?Sized,
 {
+    /// Transmutes a `Src` value into a `Self`.
+    ///
+    /// # Safety
+    ///
+    /// The safety obligations of the caller depend on the value of `ASSUME`:
+    /// - If [`ASSUME.alignment`](Assume::alignment), the caller must prove that
+    ///   all references in `src` (including `src` itself, if `Src` is a
+    ///   reference type) have well-aligned referents upon transmutation into
+    ///   `Self`.
```

Comment:
IIUC, this is assuming that the only types that may be transmuted into references are other references, and thus the set of references in `src` is a (possibly trivial) super-set of the set of references in the resulting `Self`. It might be more straightforward and more future-proof to not make this assumption and only refer directly to the set of references in `Self`.

Also, IMO the wording "well-aligned referent" is confusing since it makes it seem like it's a property of the content of the referent rather than a property of the *address* of the referent. There are different ways you could rephrase this, but I'd suggest using a phrasing that treats alignment as a property of the reference, not a property of the referent.

Reply:
Let's make further comments about this here, to keep things centralized: https://github.com/rust-lang/rust/pull/129032#discussion_r1714491421

Reply:
I worry that "well-aligned reference" suggests that the address of the reference is well-aligned, rather than the address of the referent. What do you think of:
```suggestion
    /// - If [`ASSUME.alignment`](Assume::alignment), the caller must guarantee
    ///   that references in the returned `Self` have well-aligned referents.
```

Reply:
Maybe to be slightly more verbose:

> If `ASSUME.alignment`, the caller must guarantee that the addresses of references in the returned `Self` satisfy the alignment requirements of their referent types.

Or alternatively:

> If `ASSUME.alignment`, the caller must guarantee that the addresses of references in the returned `Self` are properly aligned for their referent types.

(The phrase "properly aligned" in my mind implies a correctness property which is binary - it's either satisfied or not. By contrast, "well-aligned" feels like it's describing something on a scale, or something that is subjective. That's a tiny nit and not important.)

Reply:
[Done](https://github.com/rust-lang/rust/compare/96cfefbb125916871760c1df4c2ee29944635e90..1c0e7fac16490ea2e83ab0127c364231a42742ad#diff-85698661a7eb53801df364c9242c8613f7d2e4bca026bc5c88d542c22fa01ddfR90-R92)